### DISCUSSION TITLE
Add new board based on STM32H725AE

### DISCUSSION THREAD

**Discussion by JohnK1987** [mbed-ce/mbed-os#121] [Q&A] [answered]
Hello and happy new year,

I am trying add a STM32H725AE based board called [Wio Lite AI](https://www.seeedstudio.com/Wio-Lite-AI-Single-Board-p-5120.html) from Seed Studio.
* [Schematic](https://github.com/radiohound/SeedStudio_Wio_Lite_AI/)
* [Reference](https://github.com/Seeed-Studio/ArduinoCore-stm32/tree/master/variants/WIO_LITE_AI) (Here I took the clock settings, and I also tried that by my own, but still nothing)

I am facing an issue and i don't know what i should do.

Directly after start I received MbedOS crash
```
++ MbedOS Error Info ++
Error Status: 0x80FF0100 Code: 256 Module: 255
Error Message: Fatal Run-time error
Location: 0x8000E91
Error Value: 0x0
Current Thread: <unnamed> Id: 0x0 Entry: 0x524F3525 StackSize: 0xB4CCB67C StackMem: 0xDB97F33D SP: 0x2404FD8C 
For more info, visit: https://mbed.com/s/error?error=0x80FF0100&tgt=WioLiteAI_H725AE
-- MbedOS Error Info --
HAL_RCC_OscConfig ERROR
```
The error comes from [lp_ticker.c](https://github.com/mbed-ce/mbed-os/blob/afba4e1eb51885124396c5b46d4af30105cce625/targets/TARGET_STM/lp_ticker.c#L220). So then I disable LPTICKER in targets.json configuration, but then the board stucked directly after start of program. 

```
int main()
{
    debug_if(DEBUG, "Test\n");
    while (true) {
        //debug_if(DEBUG, "loop\n");
        led1 = !led1;
        thread_sleep_for(1000); // does not matter if the sleep is here or not ,the behavior is same
    }
}
```
Output in console is just first char T, the led switch to ON and then do nothing more, no errors, just stay.

Targets.json content for this board
```
    "WIOLITEAI_H725AE": {
        "inherits": [
            "MCU_STM32H725xE"
        ],
        "config": {
            "hse_value": {
                "help": "HSE default value is 25MHz in HAL",
                "value": "25000000",
                "macro_name": "HSE_VALUE"
            }
        },
        "device_has_remove": [
            "LPTICKER"
        ],
        "overrides": {
            "system_power_supply": "PWR_DIRECT_SMPS_SUPPLY"
        },
        "device_name": "STM32H725AEIx"
    },
```

I also tried just simple blinky with same clock settings in STM32CubeIDE and seems to be OK.

In next few days I will try to debug it, but I would like to know If i miss something.
@JojoS62 @multiplemonomials  some hints?

Thank you
BR, Jan

**Comment by multiplemonomials** [comment, marked as answer]
If I were you, I might try to step through the STM32 HAL code and see why the lpticker OscConfig call is failing.  It might help you find what's different about this board vs others that Mbed supports.

**Reply by JohnK1987** [reply]
Ok, it seems the LPTICKER is failing because LSE does not ready after 5sec - https://github.com/mbed-ce/mbed-os/blob/280d40c5c7a3c7d6c5ac58d3fb7ca31f550bf3cb/targets/TARGET_STM/TARGET_STM32H7/STM32Cube_FW/STM32H7xx_HAL_Driver/stm32h7xx_hal_rcc.c#L719-L725

I will continue tomorrow.
BR, Jan

**Reply by multiplemonomials** [reply]
Is the LSE oscillator mounted on this board?  Maybe it needs a different drive strength setting or something?

**Reply by JohnK1987** [reply]
I believe it is present. Bigger package should be HSE and smaller should be LSE.
I will check it today later and also found this https://github.com/ARMmbed/mbed-os/issues/13781

**Reply by JojoS62** [reply]
The LSE is in the schematics, it seems to be a 12 MHz crystal. Have you checked the system_clock.c? I'm just trying to find it.
lp_ticker.c has also a lot of target specific code, maybe the H725 is not handled correctly here.
And this is a interesting board, have ordered it also.

**Reply by multiplemonomials** [reply]
Looking at the code linked above, it looked like they do not enable the LSE and instead use internal clock.  FWIW, I've had some super weird issues with STM32 LSE crystals not working for no explainable reason in the past, and have sworn off them completely (since we don't need an RTC for my applications).

**Reply by JojoS62** [reply]
For the LSE xtal problem I had supplied a PR and it is merged. There is a good document from ST about suitable crystals for STM32 and the MCU are different. The most complex are the H7 where the drive load can be configured in low/mid/high driving load. Newer low power crystals must be used with low drive power, otherwise they can be damaged. Old large size crystals need more power to oszillate reliable.
The drive load can be configured in mbed_app.json

**Reply by JohnK1987** [reply]
> The LSE is in the schematics, it seems to be a 12 MHz crystal.

From my understanding is FC-12M just a marking of component type and the real value si 32.768kHz. 

Ok, I understand there some necessary settings around LSE, but I supposed is not necessary for Mbed OS basic functionality and that is why I wanted try set the LSE OFF, remove the LPTICKER and RTC. Then I hoped it will work, but not.

> Have you checked the system_clock.c? I'm just trying to find it.

I used the clock code from Seed repo (above) and also I set my own via CubeMX, but nothing seems to change the state. The RCC_LSEDRIVE_LOW for LSE is set in default for all STM families (apart from F1 and F4) in the targets.json
The rest seems to be in [TARGET_STM/mbed_overrides.c](https://github.com/mbed-ce/mbed-os/blob/280d40c5c7a3c7d6c5ac58d3fb7ca31f550bf3cb/targets/TARGET_STM/mbed_overrides.c#L63-L70)

However I am not expert so I believe I miss something.

> And this is a interesting board, have ordered it also.

Yeah, I found it when I order the second Blackpill. Seems be powerful, small with interesting equipment. Two negative things are small amount of accessible pins and that Boot0 is just on testpoint. I spend hours to find out how to put the board into DFU mode (because it was stuck and SWD was not working).

**Reply by multiplemonomials** [reply]
Hmm, if you have configured lse-available to false in JSON, you shouldn't need to also disable LPTICKER.  Disabling LSE should make the LPTICKER use the LSI oscillator.

**Reply by JojoS62** [reply]
> > The LSE is in the schematics, it seems to be a 12 MHz crystal.
> 
> From my understanding is FC-12M just a marking of component type and the real value si 32.768kHz.

You're right, there was no value for the xtal and 12M sounds like 12 MHz. The crystal is also very tiny with 0,5 µW power, so it should work with the low drive default.

**Reply by JohnK1987** [reply]
Thank you for your idea Jamie, however I already tried that last night, again with negative result, but...  I played with settings directly in STM32CubeIDE (a pure blinky project), it is little bit faster (in comparison with .json file, where is necessary re-compile whole project for every change) because all settings is in one file and necessary settings are included automatically. Also debugging is too much faster. Debugging the Mbed in VS Code is a real pain (but it can be caused by huge content of Mbed code or my low knowledge), always gets stuck somewhere and loading something for 20sec and so on. Ok, back to the issue...

I changed the settings to the LSI in STM32CubeIDE what was also not working and I found I need to do full memory erase when switch from LSE settings to LSI settings, without that the LSI also does not get ON (ready). 
With that knowledge I tried that again with Mbed. The behavior was changed, but not to a positive state. The board just did nothing :D
So I ran Mbed debug again and found the code stucks in a function _SCB_EnableDCache()_ called from[TARGET_STM/mbed_overrides.c - Line 172 ](https://github.com/mbed-ce/mbed-os/blob/280d40c5c7a3c7d6c5ac58d3fb7ca31f550bf3cb/targets/TARGET_STM/mbed_overrides.c#L171-L173). That was something new, I commented out this code and tadáááá, the basic code (with RTC and LPTICKER enabled and drived by LSI) from above is working as expected!

Now I want to understand,  why it is failing with _SCB_EnableDCache()_ on this target and try to understand the LSE issue. How I saw on some web discussions, it seems the LSE is very sensitive to the PCB design and also capacitors, but I hope it is not this case.

I will try understand the Dcache issue, and then I will perform the Greentea .

BR, Jan

**Reply by JohnK1987** [reply]
I found only this until now. The _SCB_EnableDCache()_ function contains a loop what should invalidate D-Cache and it is same implementation like in _SCB_InvalidateDCache()_ funcion. So the real problem is not Enable but Invalidate.

I also tried Greentea and the result was not so bad, two tests below and USB tests failed. All USB APIs failed on same point - assertion of PCD and FIFO configuration.

https://github.com/mbed-ce/mbed-os/blob/280d40c5c7a3c7d6c5ac58d3fb7ca31f550bf3cb/targets/TARGET_STM/USBPhy_STM32.cpp#L341-L347

```
97% tests passed, 2 tests failed out of 71

Total Test time (real) = 604.15 sec

The following tests did not run:
         14 - mbed-hal-ospi (Skipped)
         16 - mbed-hal-qspi (Skipped)
         32 - mbed-hal-verify-arduino-pinmap (Skipped)
         43 - mbed-platform-crash-reporting (Skipped)
         53 - mbed-platform-stats-cpu (Skipped)
         54 - mbed-platform-stats-heap (Skipped)
         55 - mbed-platform-stats-sys (Skipped)
         56 - mbed-platform-stats-thread (Skipped)
         71 - mbed-connectivity-nfc-eeprom (Skipped)

The following tests FAILED:
         13 - mbed-hal-mpu (Failed)
         23 - mbed-hal-sleep-manager (Failed)
Errors while running CTest
```
BR, Jan

**Reply by JohnK1987** [reply]
I asked Jerome about the mentioned problems.
About LSE: _My issue is, I think, that all mbed-os power features are based on LSE support on ST side…
We used to disable LPTICKER in that case…_

**Reply by JohnK1987** [reply]
OK, the USB issue were my stupidity and low knowledge. How I tried different settings of system_power_supply I forgot change them back to right one. USB is working now.

So the last is - SCB_InvalidateDCache()...

**Reply by JohnK1987** [reply]
Good news! After some pause with Blackpill, I went back to this one and I finally found a rootcause of behavior descripted above.

In the past I tried debugging the unexpected behavior and I found the ST-Link V2.1 stucked in function what was responsible for Dcache enable. After short test, where I commented out this method, the wrong behavior was gone. 
However it seems it was just a luck because I was wrong :) I bought STLINK-V3MINIE, which is more pleasant to work with, and I found the code stucks much later.
The issue is wrong SRAM memory settings in cmsis_nvic.h.
```
#if !defined(MBED_RAM_START)
#define MBED_RAM_START  0x24000000
#endif

#if !defined(MBED_RAM_SIZE)
// 0x38000000 - 0x38003FFF 16K SRAM4
// 0x30004000 - 0x30007FFF 16K SRAM2
// 0x30000000 - 0x30003FFF 16K SRAM1
// 0x24000000 - 0x2404FFFF 320K AXI SRAM
// 0x20000000 - 0x2001FFFF 128K DTCM
#define MBED_RAM_SIZE  0x50000  // 320 KB
#endif
```
I do not understand all context, but I found on other targets (H723, H743) the DTCM is used like standard RAM and the AXI is used like RAM1. So I set it also like that and tadááá, the issue is gone!

```
// 0x38000000 - 0x38003FFF 16K SRAM4
// 0x30004000 - 0x30007FFF 16K SRAM2
// 0x30000000 - 0x30003FFF 16K SRAM1
// 0x24000000 - 0x2404FFFF 320K AXI SRAM
// 0x20000000 - 0x2001FFFF 128K DTCM

#if !defined(MBED_RAM_START)
#define MBED_RAM_START  0x20000000
#endif

#if !defined(MBED_RAM_SIZE)
#define MBED_RAM_SIZE  0x20000  // 128 KB
#endif

#if !defined(MBED_RAM1_START)
#define MBED_RAM1_START  0x24000000
#endif

#if !defined(MBED_RAM1_SIZE)
#define MBED_RAM1_SIZE  0x50000  // 320 KB
#endif
```

Failed Greentee tests remain same for now.

BR, Jan

**Reply by JojoS62** [reply]
Ram/Rom sizes should be checked in the datasheet when in doubt. My recent problem was also a wrong Ram setting. These are imported from Keil pack manager, the large index.json that was in mbed-os in tools. This file is not bug free, as you can see in my PR in mbed-os for this issue.
DCache and DMA is also difficult and very different between the MCUs. DMA is not supported in some RAM regions, and this can be only solved by modifying the linker script and adding sections to the code. That is why the custom linker script option was very important in my opinion. There is also the MPU involved for turning of Cache on/off only for some regions instead of disabling all.
But here I see also missing OS support, the MPU entries should be handled by OS functions. E.g., the Ethernet code disables the complete DCache when Ethernet is used (on H7).
A clean solution takes time to develop and to test, especially when you want to support more than only STM32 targets.

**Reply by JohnK1987** [reply]
Yeah @JojoS62, I check the datasheet before of course, but I do not understand these things around linker script/memory mapping. 
> DCache and DMA is also difficult and very different between the MCUs. DMA is not supported in some RAM regions, and this can be only solved by modifying the linker script and adding sections to the code.

That seems be the information what I needed in January :) For now I solve that with settings (linker script/memory mapping) from STM32H723, but I am not sure if this is 100% correct.

Another good news is the LSE is up. It was **lse_drive_load_level** settings how was expected, but I did not know I need to do complete power down between change of this setting. 
During testing with STM32CubeIDE I just did a change of this and upload to the board, reset, test. When it was not working, then changed it to another one and so on. Yesterday I found I need to unplug USB cable between these actions... sad.

However, the Greentea test **mbed-hal-sleep-manager** still failing with assertion - [lp ticker sleep time incorrect](https://github.com/mbed-ce/mbed-os/blob/2d7f24285093e0c5135e5518a4d8ed54ff33fd0f/hal/tests/TESTS/mbed_hal/sleep_manager/main.cpp#L183)

BR, Jan

**Reply by multiplemonomials** [reply]
> I did not know I need to do complete power down between change of this setting.

Waaaat?  Dang that might explain a similar issue I ran into in the past, and ended up giving up on.

**Reply by JohnK1987** [reply]
Yeah, it is annoying.

In the context of issues on this board I found this
* the issue with **mbed-hal-sleep-manager** is not target specific, but STM32H7 family specific - https://github.com/mbed-ce/mbed-os/issues/131
* the issue with mbed-usb-device-basic  is related to STM32 in general - https://github.com/mbed-ce/mbed-os/issues/53

So it seems I can ignore these issues, for now and make a PR - https://github.com/mbed-ce/mbed-os/pull/133

BR, Jan
